Support translations of certain math notation

danielhollas opened the pull request:
This PR adds a functionality for translations of certain math notations for different locales, as specified in this [google doc](https://docs.google.com/spreadsheets/d/1qgi-KjumcZ6yru19U5weqZK9TosRlTdLZqbXbABBJoQ/edit#gid=0).

Only the most obvious cases from that doc are covered here: 
1. Decimal points to decimal commas
2. multiplication and division signs
3. latin version of sinus operator

Actually, the functionality was partially already in the repo, albeit not working. I fixed it, added the cases mentioned above and added some tests. Note that even this partial inelegant solution will be a BIG help for translators.

This PR partially solves #5 and is tracked in [JIRA ticket IC-116](https://khanacademy.atlassian.net/browse/IC-116).

Example: String 
`The answer is $5.6$`
need to be translated to Czech as 
`Výsledek je $5{,}6$`
But the previous version would detect the change in decimal point as a change in math code and would not offer Smart Translations.

**TODO before merge**
1. Ask the new advocates to fill in the google doc and update the code.
2. Think about the default behaviour when math notation in the translated string does not match our expectation. In the current version, I decided to be conservative and in the case of mismatch simply do not offer translations. For example, if the user forgets to translate the decimal point, we will not offer smart translations, which may potentially notify him that something is wrong.

danielhollas commented:
Note that in order to compile the code with babel, I had to turn off the `--loose` option in package.json. I actually committed this change, which might have been a bad idea. :slightly_frowning_face: 

If this causes problems, somebody else will have to do the build.

mpolyak commented:
> Note that in order to compile the code with babel, I had to turn off the --loose option in package.json. I actually committed this change, which might have been a bad idea.

@danielhollas, yes lets not commit this change.

Last time I had to rebuild after merging a pull request in order for it to work with Translations Editor, so that's not a problem and I wouldn't worry about it, as long as you can run the JS tests for your pull request.

danielhollas commented:
Okay, I put the loose option back and did a proper build, see my last comment at #7.

danielhollas commented:
Hi!
Just so I understand, you added these tests and they are failing?

When I added them, all of them passed except "fr doesn't work with '\\text". I am looking into it...

danielhollas commented:
Okay, so it turns out that we need to translate Math inside `getMathDictionary()` as well. I will push the fix shortly. Thank you for the find! I should have tested that...

mpolyak commented on src/translation-assistant.js:
```diff
@@ -23,6 +23,7 @@ const WIDGET_REGEX = /\[\[[\u2603][^\]]+\]\]/g;
 
 // TODO(michaelpolyak): Add support for other \text commands:
 // https://github.com/Khan/KaTeX/blob/3280652bd68973ad9edd73273137049324c5cab9/src/functions.js#L50
+// Comment(danhollas): Those other commands are rare/non-existent in KA corpus
```
Can you change this to NOTE(danhollas): ... please?

mpolyak commented on src/translation-assistant.js:
```diff
@@ -129,16 +130,23 @@ function getMapping(
     const outputs = translatedStr.match(findRegex) || [];
 
     if (findRegex === MATH_REGEX) {
+        inputs = inputs.map(
```
nit: You could chain the two maps as follows:

```
        inputs = inputs
            .map((input) => translateMath(input, lang))
            .map((input) => replaceTextInMath(input, mathDictionary));
```

mpolyak commented on src/translation-assistant.js:
```diff
@@ -129,16 +130,23 @@ function getMapping(
     const mapping = [];
 
     outputs.forEach((output, outputIndex) => {
+        /*
```
Lets change inline comment back to `//` please.

mpolyak commented on src/translation-assistant.js:
```diff
@@ -348,20 +361,42 @@ function createTemplate(englishStr, translatedStr, lang) {
 
 /**
  * Handles any per language special case translations, e.g. Portuguese uses
```
Lets format the comment:

```
/**
 * Handles any per language special case translations, e.g. Portuguese uses
 * `sen` instead of `sin`.
 *
 * Many languages use decimal comma, according to this table:
 * https://docs.google.com/spreadsheets/d/1qgi-KjumcZ6yru19U5weqZK9TosRlTdLZqbXbABBJoQ/edit#gid=0
 *
 * TODO(danielhollas): For now only the obvious cases were implemented.
 * TODO(danielhollas): Need to update this when new langs join translations.
 *
 * @param {string} math A math expression to translate for locale.
 * @param {string} lang The locale of the translation language.
 * @returns {string} The translated math expression.
 */
```

mpolyak commented on src/translation-assistant.js:
```diff
@@ -348,20 +361,42 @@ function createTemplate(englishStr, translatedStr, lang) {
-        return math;
-    }
+
+    const mathTranslations = [
```
Very cool!

mpolyak commented on tests/assistant_spec.js:
```diff
@@ -478,6 +479,182 @@ describe('TranslationAssistant (math)', function() {
+        ], 'cs');
+    });
+
+    // TODO(danielhollas): This we need to decide
```
Do you wish to come to a decision before merging this pull request?

If so maybe ping Caroline in JIRA? Otherwise, if you're ok addressing this at a later point, as a separate change, lets expand the comment a bit to provide more context.

danielhollas replied:
I've given it a lot of thought and feel pretty good about the current behavior. I will delete this TODO and only leave a NOTE in the source. After this is deployed, I'll ask the i18n community for feedback and change this behavior if needed.

mpolyak commented on tests/assistant_spec.js:
```diff
@@ -478,6 +479,182 @@ describe('TranslationAssistant (math)', function() {
     });
 });
 
+describe('TranslationAssistant (math-translate)', function() {
```
Thank you for all these tests!

mpolyak commented on tests/assistant_spec.js:
```diff
@@ -62,7 +62,7 @@ const graphieLink6 = makeGraphieLink();
 const getEnglishStr = (item) => item.englishStr;
 const getTranslation = (item) => item.translatedStr;
-const lang = 'fr';
+const lang = 'cs';
```
Since this is the default, how about we handle it within `assertSuggestions` itself, and remove this file level const?

If we change `assertSuggestions` signature to:

```
function assertSuggestions(allItems, itemsToTranslate, translatedStrs,
        lang = 'cs') {
```

Then in any test case that's currently using this default language we can simply omit passing in `lang` into `assertSuggestions` so it defaults to `cs`?

danielhollas commented:
Thank you for the review! Hopefully, I addressed all the points. 

I added a link to the example in the Test Plan.

I would suggest to merge and deploy this. I will then make the announcement to the advocate community and ask for feedback. I'll then submit the following PR, which will hopefully only contain updated locale lists in the `mathTranslations` dictionary. Is that okay with you?

danielhollas commented:
Oh, one more important thing. I restricted link sharing access on that Google doc to only people at KA and invited people. Up to now, anyone with that link could edit. :-( I've already mentioned this link several times here on Github, in the source code and I put it in the summary commit message. 

Alternatively, I can go back and try to delete the links, but this seemed like a safer option. What do you think? The downside to this is that from now we will have to explicitly invite every advocate...:-/

mpolyak commented:
> I added a link to the example in the Test Plan.

Thank you!

> I would suggest to merge and deploy this. I will then make the announcement to the advocate community and ask for feedback.

I see the TODO still in the tests, did you mean to change that to a NOTE?

> I'll then submit the following PR, which will hopefully only contain updated locale lists in the mathTranslations dictionary. Is that okay with you?

Sounds like a plan 👍 

> I restricted link sharing access on that Google doc to only people at KA and invited people.

I think that should be ok. You could also allow suggestions or comments to have new people contribute? In any case we'd need to track and approve adding translations, so I don't see it as an issue.

danielhollas commented:
I see the TODO still in the tests, did you mean to change that to a NOTE?

Whoops, amended, sorry.

mpolyak commented:
Thanks @danielhollas!

![seal-of-approval](https://user-images.githubusercontent.com/4722598/43090606-ddf599ae-8e75-11e8-968d-9b70b9ead439.png)
